core/dbt/context/common.py

```diff
@@ -250,25 +249,12 @@ def assert_var_defined(self, var_name, default):
                 self.model
             )
 
-    def assert_var_not_none(self, var_name):
-        raw = self.local_vars[var_name]
-        if raw is None:
-            pretty_vars = self.pretty_dict(self.local_vars)
-            dbt.exceptions.raise_compiler_error(
-                self.NoneVarError.format(
-                    var_name, self.model_name, pretty_vars
-                ),
-                self.model
-            )
-
-    def __call__(self, var_name, default=None):
+    def __call__(self, var_name, default=_VAR_NOTSET):
```

Comment:
As I understand it, supplying an object (or non-primitive) as a default argument means that Python uses the pointer to that object as the default value for all invocations of a method. This is the thing you're taking advantage of here, right?

This might be way off the mark, but... is this... thread-safe? Is `_VAR_NOTSET` initialized when the `common.py` file is imported? This is a neat construct!

Reply:
Mostly! Python actually guarantees that supplying _any_ value as a default argument means that Python uses that object as the default value. The tricky part that confuses people is that there is in fact only one each of `None`, `True`, and `False`. The `id()` of an object is guaranteed to be unique, and `a is b` basically checks `id(a) == id(b)`.

This is a somewhat common technique for handling this irritating case of wanting to distinguish between `None` and "not set" on an optional arg. The other way I've seen is doing `**kwargs` and checking what's in there, which I'm not really fond of.

I'm not sure where I first stumbled across it, but I have been using it for a long time and I sure didn't come up with it on my own. Because it's the default argument, the user can never intentionally pass it, so it can only arrive on a function signature. I don't think it's possible to go wrong here from a thread/process safety perspective. You probably shouldn't subclass and override this method or the value though!

If you need even more comfort: Because the variable is defined at the class level, and the class is defined at the module level, the identity of the object is defined at module import time. Python guarantees that module loading is thread-safe by having a lock around imports, so you know that this value will only be set once and be visible to all threads.